lib/spack/spack/test/zz/bindist_def.py

```diff
@@ -0,0 +1,416 @@
+                                  pspec.name])
+        buildcache.buildcache(parser, args)
+
+    # list the buildcaches in the mirror
+    create_args.insert(create_args.index('-a'), '-f')
+    args = parser.parse_args(create_args)
+    buildcache.buildcache(parser, args)
+    args = parser.parse_args(['list', '-f', '-l', '-v'])
+    buildcache.buildcache(parser, args)
+
+    # Uninstall the package and deps
+    uparser = argparse.ArgumentParser()
+    uninstall.setup_parser(uparser)
+    uargs = uparser.parse_args(['-y', '--dependents', zspec.name])
+    uninstall.uninstall(uparser, uargs)
+
+    # test install
+    args = parser.parse_args(install_args)
+    buildcache.buildcache(parser, args)
+
+    # This gives warning that spec is already installed
+    buildcache.buildcache(parser, args)
+
+    # test overwrite install
+    install_args.insert(install_args.index('-a'), '-f')
+    args = parser.parse_args(install_args)
+    buildcache.buildcache(parser, args)
+
+    # Add a second key
+    if has_gnupg2():
+        gparser = argparse.ArgumentParser()
+        gpg.setup_parser(gparser)
+        gargs = gparser.parse_args(['create', 'test key 2',
+                                    'spack@googlegroups.com',
+                                    '--expires', '0', '--comment',
+                                    '"Spack test key1"',
+                                    '--export', '%s/build_cache/test2.key' %
+                                    str(mirror_path_def)])
+        gpg.gpg(gparser, gargs)
+
+        create_args.insert(create_args.index('-a'), '-f')
+        args = parser.parse_args(create_args)
+        # Trigger pick key warning
+        # buildcache.buildcache(parser, args)
+
+    args = parser.parse_args(['keys', '-f'])
+    buildcache.buildcache(parser, args)
+
+    args = parser.parse_args(['list'])
+    buildcache.buildcache(parser, args)
+
+    args = parser.parse_args(['list', '-f'])
+    buildcache.buildcache(parser, args)
+
+    args = parser.parse_args(['list', '-l', '-v'])
+    buildcache.buildcache(parser, args)
+    bindist._cached_specs = None
+    spack.stage.purge()
+    shutil.rmtree(os.path.join(mock_stage, 'build_cache'))
+
+
+@pytest.mark.maybeslow
+@pytest.mark.usefixtures('gpg_directory', 'default_config',
+                         'install_dir_non_default_layout', 'cache_directory')
+def test_install_def_nondef_def(tmpdir, mock_stage):
+    """
+    Test the creation and installation of buildcaches with default rpaths
+    into the non-default directory layout scheme.
+    """
+    cparser = argparse.ArgumentParser()
+    compiler.setup_parser(cparser)
+    cargs = cparser.parse_args(['add', '--scope', 'site'])
+    compiler.compiler(cparser, cargs)
+    cargs = cparser.parse_args(['list', '--scope', 'site'])
+    compiler.compiler(cparser, cargs)
+    cargs = cparser.parse_args(['info', 'clang'])
+    compiler.compiler(cparser, cargs)
+    cargs = cparser.parse_args(['info', 'gcc'])
```

Comment:
What is the purpose of this section?

Reply:
I needed to add a real (system) compiler for the build tests.